crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,238 @@
-use std::cell::RefCell;
-use std::fs;
+use std::collections::HashMap;
+use std::fs::{self, File};
 use std::hash::Hasher;
-use std::io::Write;
-#[cfg(unix)]
-use std::os::unix::fs::PermissionsExt;
-use std::path::Path;
+use std::io::{self, BufReader, BufWriter, Write};
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
+use std::time::SystemTime;
 
-use anyhow::Result;
-use filetime::FileTime;
-use log::error;
-use path_absolutize::Absolutize;
-use ruff_text_size::{TextRange, TextSize};
-use serde::ser::{SerializeSeq, SerializeStruct};
-use serde::{Deserialize, Serialize, Serializer};
+use anyhow::{anyhow, Context, Result};
+use serde::{Deserialize, Serialize};
 
 use ruff::message::Message;
-use ruff::settings::{AllSettings, Settings};
+use ruff::settings::Settings;
 use ruff_cache::{CacheKey, CacheKeyHasher};
 use ruff_diagnostics::{DiagnosticKind, Fix};
 use ruff_python_ast::imports::ImportMap;
 use ruff_python_ast::source_code::SourceFileBuilder;
+use ruff_text_size::{TextRange, TextSize};
 
-const CARGO_PKG_VERSION: &str = env!("CARGO_PKG_VERSION");
-
-/// Vec storing all source files. The tuple is (filename, source code).
-type Files<'a> = Vec<(&'a str, &'a str)>;
-type FilesBuf = Vec<(String, String)>;
-
-struct CheckResultRef<'a> {
-    imports: &'a ImportMap,
-    messages: &'a [Message],
-}
-
-impl Serialize for CheckResultRef<'_> {
-    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let mut s = serializer.serialize_struct("CheckResultRef", 3)?;
-
-        s.serialize_field("imports", &self.imports)?;
-
-        let serialize_messages = SerializeMessages {
-            messages: self.messages,
-            files: RefCell::default(),
+use crate::diagnostics::Diagnostics;
+
+/// On disk representation of a cache of a package.
+#[derive(Deserialize, Debug, Serialize)]
+pub(crate) struct PackageCache {
+    /// Location of the cache.
+    ///
+    /// Not stored on disk, just used as a storage location.
+    #[serde(skip)]
+    path: PathBuf,
+    /// Path to the root of the package.
+    ///
+    /// Usually this is a directory, but it can also be a single file in case of
+    /// single file "packages", e.g. scripts.
+    package_root: PathBuf,
+    /// Mapping of source file path to it's cached data.
+    // TODO: look into concurrent hashmap or similar instead of a mutex.
+    files: Mutex<HashMap<RelativePathBuf, FileCache>>,
+}
+
+impl PackageCache {
+    /// Open or create a new package cache.
+    pub(crate) fn open(
```

Comment:
I think this approach solves a problem that exists with the cache today, but I'd like to confirm...

Today, we use the cache key as the filename. When file contents are invalidated, we write a new file to the cache (since the cache key has changed), but leave the old file around. So the cache, as it exists today, grows monotonically, and gets bigger and bigger over time.

As far as I can tell, this new approach will replace the cache entirely when contents change, so we're effectively bounding the size of the cache by the size of the package itself. Is that right? If so, it should completed solve the problem of the cache growing eternally as seen in #5132.

Reply:
(The benefit of the current approach is that if you change a setting, then revert that change, we still have the cached results from the previous settings. But that's so rare as to be not at all worthwhile.)

Reply:
> As far as I can tell, this new approach will replace the cache entirely when contents change, so we're effectively bounding the size of the cache by the size of the package itself. Is that right? If so, it should completed solve the problem of the cache growing eternally as seen in #5132.

It almost solves it, but there are some limitations. This design doesn't remove the `files` hashmap entry for a source file that has been (re)moved. We could solve this my not reusing the read cache and instead recreate the `files` hashmap before storing it.

The second limitation is the hashkey generation. Because it uses various settings to ensure the cache is valid it doesn't actually remove old cache files if, e.g., a project changes the settings. I think this can be solved by splitting the file name into `$project_hash.$every_thing_else_hash` this way we can do a `rm $project_hash.*` periodically or something. But as discussed yesterday for now we can use `ruff clean` to "solve" both problems (by removing all cache files for the project, or after https://github.com/astral-sh/ruff/pull/5122 all cache files for all projects).